csrc/id_model/indexing.cpp

```diff
@@ -120,6 +137,20 @@ std::tuple<std::vector<IterDomain*>, std::vector<Val*>> getAllocationDomains(
     }
   }
 
+  // Loop promotion may affect allocations. Promotions of intermediate
+  // domains may not be defined correctly. Only consider loop domains
+  // for now.
```

Comment:
![image](https://github.com/NVIDIA/Fuser/assets/1032377/99efbcbd-cba7-4b39-b070-480592722d51)
What if we have the above schedule? How should we handle this? Should we just raise an error, or allocate the tv with broadcasting as `I0*I1`, or allocating it as `5`?

Reply:
When the allocation domain is not the loop domain, we don't have any logic other than fully allocating `bxI1`. If we want to just allocate `5`, we could do it by setting the allocation domain as the loop domain.

I wonder what the domains and parallelization would look like with TMA. 

I believe this is more of a question on what the allocation domain should be. I think that ideally `getAllocationDomains` here should be just a trivial function call to `tv->getAllocationDomain` and each tensor should always have its correct allocation domains.

Reply:
> I believe this is more of a question on what the allocation domain should be.

I agree. I think we need a restriction that, each ID in the allocation domain must be either fully inlined or fully not inlined. It can not have an ID coming from a merge of an inlined ID with an uninlined ID, or an ID who is the parent of a split whose outer is inlined but inner not.

> I wonder what the domains and parallelization would look like with TMA.

In my mental model, it is similar to above: IDs in the allocation domain must be either a tile or a non-tile, it can not be a mix of both. However, in practice, for now, even if we have a mix, the code should still work (thanks to some existing hack in our system? and the same hack will also make the above example just allocate 5 today?)

Reply:
> > I believe this is more of a question on what the allocation domain should be.
> 
> I agree. I think we need a restriction that, each ID in the allocation domain must be either fully inlined or fully not inlined. It can not have an ID coming from a merge of an inlined ID with an uninlined ID, or an ID who is the parent of a split whose outer is inlined but inner not.

Yeah, I think you could say that the mechanism of promotion is making partially inlined domains fully inlined. In the above case, the innermost `5` domain of the broadcast tensor is promoted to the innermost `5` domain of the non-broadcast tensor, meaning it's effectively fully inlined.

I think we have a reasonable understanding of promotions of loop domains. Can we propagate promotions to allocation domains that are not between logical and loop? I guess we also have a similar problem of propagating parallel types from loop domains to allocation domains.
> 
> > I wonder what the domains and parallelization would look like with TMA.
> 
> In my mental model, it is similar to above: IDs in the allocation domain must be either a tile or a non-tile, it can not be a mix of both. However, in practice, for now, even if we have a mix, the code should still work (thanks to some existing hack in our system? and the same hack will also make the above example just allocate 5 today?)

As long as the allocation domain of the tensor is just the loop domain, indexing is trivial thanks to the loop promotion. How it's implemented in the current main branch isn't that different conceptually, but the implementation is quite convoluted since, for example, it always traverses back to logical domains, whereas in this case we can just directly index the loop (=allocation) domains.